### PR TITLE
Turn data-tiling on by default.

### PR DESCRIPTION
**Pull request by hanhanW**
The revision turns data-tiling ons by default. In the commit:

- Disable data-tiling for transform dialect tests
- Allow tile size adjustment failing if it is going to be on vmvx ukernel path. All the inner tile sizes are `kDynamic`, and we have nothing to do for adjustment.
- Move all related passes under `transformOptions.options.dataTiling` control.

We can get up to 10x faster for models haven't been over-looked (e.g., GPT2, Falcon7bGptqPT, BertLarge, etc); we have 25%, 36%, 45% slow down for some models we've been working on previously. The full benchmark result (with microkernels on) can be found at https://gist.github.com/iree-github-actions-bot/b2ac57f9f1fd6894e93da19d6137c5da

### CODE REVIEW

**Comment by hanhanW** [comment]
The biggest regression on llvm-cpu side is DeepLabV3, MobileBert, MobileNetV2, and MobileSSD on 8-threads. This is a known issue about distribution.

For single-threaded regression, they are 7-20% -- which is acceptable to me. See the below snapshot for highlights:

image


I don't really care about vmvx regression, they are likely because of batch_matmul issues. If the microkernel is on, data-tiling is not applied to batch_matmul. In the context, the regression is expected. It's been tracking in https://github.com/openxla/iree/issues/15314

**Comment by dcaballe** [comment]
> The biggest regression on llvm-cpu side is DeepLabV3, MobileBert, MobileNetV2, and MobileSSD on 8-threads. This is a known issue about distribution.

Is this because DT is using different distribution sizes to non-DT?

**Comment by hanhanW** [comment]
> Is this because DT is using different distribution sizes to non-DT?

DT introduces mmt4d ops, pack ops, and unpack ops, while non-DT does not have those ops. The mmt4d distribution tile sizes could be outdated, which was implemented 1 to 2 years ago by me. The pack/unpack distribution tile sizes are not tuned yet. We do observe similar issue in https://github.com/openxla/iree/issues/15349#issuecomment-1789850573

**Comment by hanhanW** [comment]
@MaheshRavishankar @dcaballe @banach-space @bjacob this is ready for review. We already landed a sequence of PRs that helps functionality (https://github.com/openxla/iree/pull/15450, https://github.com/openxla/iree/pull/15453, https://github.com/openxla/iree/pull/15455, etc) and performance (https://github.com/openxla/iree/pull/15414, https://github.com/openxla/iree/pull/15397, https://github.com/openxla/iree/pull/15447, https://github.com/openxla/iree/pull/15448, https://github.com/openxla/iree/pull/15465,etc). 

We now allow backends to undo set_encoding, so it does not impact backends that haven't implemented data-tiling yet. E.g., it is a nop if the target CPU is RISC-V or Arm SVE. See https://github.com/openxla/iree/pull/15450

**Comment by hanhanW** [comment]
cc @Max191 @pzread @NatashaKnk @mariecwhite @bviyer who might be interested in this.

**Comment by hanhanW** [comment]
I will disable ukernel before landing the PR. It is mostly for getting benchmark results.

File: compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp
Comment (bjacob) [inline compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: Can you split out all logic changes like this from the actual data-tiling-by-default flip, which should ideally be a 1-line PR?

Comment (hanhanW) [reply compiler/src/iree/compiler/Codegen/LLVMCPU/KernelDispatch.cpp]: okay, I will do it!

**Comment by hanhanW** [comment]
@bjacob it is `+8 -1` now. I can create a PR that explicitly disables `data-tiling` for those tests to make this PR be `+1 -0` if you think that's better.

**Comment by bjacob** [comment]
Thanks for the splitting already, much better. Now I still think it would be best to go all the way to a 1 line change, like you describe.

Also, the current diff still turns on ukernels as well.

Thanks for all the work. I think it would be wise to wait until Monday anyway to flip the switch!

**Comment by hanhanW** [comment]
ukernels is mostly for performance check. I will disable it before land the PR. I plan to wait for a more couple days, and see if others are okay with the flip. I will make this be one-line change.

**Comment by hanhanW** [comment]
the PR is ready: https://github.com/openxla/iree/pull/15557

**Comment by dcaballe** [comment]
Posted new perf results in https://github.com/openxla/iree/issues/15313

I also see some big regressions in CI. Has anybody looked into those?

**Comment by bjacob** [comment]
I have lost track of the discussions about remaining regressions. The latest benchmark results still report some 2x regressions. Can you link to the latest thinking about them?

**Comment by hanhanW** [comment]
> Posted new perf results in #15313
> 
> I also see some big regressions in CI. Has anybody looked into those?

The big regressions happen when ukernels are disabled. There is known issue about codegen i8.i8.i32 mmt4d ops. The vector dialect and LLVM backend is not good at i8.i8.i32 things. The benchmark with ukernels on can be found at https://gist.github.com/iree-github-actions-bot/b2ac57f9f1fd6894e93da19d6137c5da and the regressions are okay

**Review by bjacob — APPROVED** [review]
Thanks for pushing it this far.  This is now a 1-line change that's a clear improvement in IREE's benchmarks.  OK to merge this pending checking with @dcaballe , but the fact that known regressions are due to non-ukernel mmt4d codegen means to me that it's time to flip this switch and deal with imperfect mmt4d codegen separately.

On a parallel front, I am working to make it easy to switch on ukernel for mmt4d selectively (as opposed to wholesale for all ops for which we might have more ukernels in the future). The first step was https://github.com/openxla/iree/pull/15571 and I'll advance that further today. So that provides another layer of reasoning why it's OK to move ahead without letting mmt4d codegen block this.

**Comment by bjacob** [comment]
Just to make this concrete, https://github.com/openxla/iree/pull/15586 shows how a 1-line change can now enable UK for mmt4d by default (without this being a UK blank cheque wrt future UK lowerings). This is saying, if the last thing preventing DT-by-default is currently suboptimal codegen for mmt4d ops, we can enable UK for mmt4d so that doesn't affect the default path anymore.

**Review by MaheshRavishankar — APPROVED** [review]
Could we merge this change with @bjacob s change to make mmt4d use ukernels and land those together. Might help with regressions (I am treating mmt4d + non-ukernel as the non-default path).

**Comment by bjacob** [comment]
Then let's discuss both in the "Mai-Tai sync" meeting in 30 min from now.  And if we confirm we want to do this now, let's merge https://github.com/openxla/iree/pull/15586 first, then this, so they're still separate (useful for bisection, we never know) and so that order ensures no intermediate state where the performance of non-ukernel mmt4d codegen is felt by default.

**Comment by MaheshRavishankar** [comment]
> Then let's discuss both in the "Mai-Tai sync" meeting in 30 min from now. And if we confirm we want to do this now, let's merge #15586 first, then this, so they're still separate (useful for bisection, we never know) and so that order ensures no intermediate state where the performance of non-ukernel mmt4d codegen is felt by default.

That sounds good to me!

**Comment by bjacob** [comment]
This has been on Discord with @dcaballe here: https://discord.com/channels/689900678990135345/1146173056537079919/1174043054282391654

The summary is that what is blocking this now is actually https://github.com/openxla/iree/issues/15242  . It's orthogonal to `mmt4d` codegen and ukernels. It's being actively worked on now by @NatashaKnk and discussed on Discord below the above linked message.

I think it is valuable to wait a bit to give this a chance to be resolved first, so we can hope for DT enablement to be a 100% clean win :-)

**Comment by hanhanW** [comment]
> Posted new perf results in #15313
> 
> I also see some big regressions in CI. Has anybody looked into those?

The regression are mostly from i8.i8.i32 mmt4d kernels. We are not generating VNNI. Issue filed: https://github.com/openxla/iree/issues/15611

**Comment by stellaraccident** [comment]
I just touched base with some of the stakeholders and it sounds like there was consensus to enable this by default. @jpienaar concur?

**Comment by hanhanW** [comment]
There is a failure after rebasing on ToT. The vmvx issue will be fixed by https://github.com/openxla/iree/pull/15705

**Comment by hanhanW** [comment]
Landing the PR because the last major compilation issue (https://github.com/openxla/iree/issues/15692) is fixed.

**Pull request merged**
